File: word_generator.py
from pyperclip import copy
import json


def simplify(word):
    word = word.replace("caca", "");
    while len(word) > 1:
        first, second = word[:2]
        if first == second:
            word = first + word[2:]
        else:
            break
    return word


def generate_words(max_length, chars="abc"):
    if (max_length > 23):
        raise Exception("nah")
    
    last_layer = list(chars)
    results = list(chars)
    for k in range(2, max_length + 1):
        layer = []
        for char in chars:
            for word in last_layer:
                new_word = char + word
                reduced = simplify(new_word)
                if (len(reduced) == len(new_word)):
                    results.append(reduced)
                    layer.append(reduced)
        last_layer = layer[:]
    
    results = ["c" + word for word in results]
    results.append("c")
    return len(results), results

def make_json(words):
    result = dict()
    for word in words:
        length = len(word)
        if result.get(length):
            result[length].append(word)
        else:
            result[length] = [word]
    return json.dumps(result)

# https://oeis.org/A068156
if __name__ == "__main__":
    length, words = generate_words(16)
    print(length)
    copy(make_json(words))

File: word_generator.js



// function simplify(word, p, q) {
//     word = word.replaceAll("caca", "").replaceAll("acac", ""); // (r3r2)^2 = 1
//     word = word.replaceAll("cb".repeat(q), "").replaceAll("bc".repeat(q), ""); // (r2r3)^q = 1
//     word = word.replaceAll("ab".repeat(p), "").replaceAll("ba".repeat(p), ""); // (r2r1)^p = 1
//     while (word.length > 1) { // r1^2 = r2^2 = r3^2 = 1
//         let first = word[0], second = word[1];
//         if (first === second) {
//             word = first.concat(word.slice(2, word.length));
//         } else {
//             break;
//         }
//     }
//     return word;
// }

function simplify(word, p, q) {
    word = word.replaceAll("caca", "").replaceAll("acac", ""); // (r3r2)^2 = 1
    word = word.replaceAll("f".repeat(q), "");
    word = word.replaceAll("r".repeat(p), "");
    while (word.length > 1) { // r1^2 = r2^2 = r3^2 = 1
        let first = word[0], second = word[1];
        if ((first === "c" || first === "a") && first === second) {
            word = first.concat(word.slice(2, word.length));
        } else {
            break;
        }
    }
    return word;
}

function generateWords(maxLength, chars, p=5, q=5) {
    const results = {1: chars.split("")};
    for (let k=2; k<=maxLength; k++) {
        results[k] = [];
        for (let chr of chars) {
            for (let word of results[k-1].slice()) {
                const newWord = chr.concat(word);
                const reduced = simplify(newWord, p, q);
                if (reduced.length == newWord.length) {
                    results[k].push(newWord);                    
                }
            }
        }
    }

    let totalWords = 0;
    for (let i=1; i<=maxLength; i++) {
        totalWords += results[i].length;
    }
    return {
        numWords: totalWords,
        words: results
    };
}